src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/DataAnnotationValidateOptions.cs

```diff
@@ -33,6 +34,7 @@ public DataAnnotationValidateOptions(string name)
         /// <param name="name">The name of the options instance being validated.</param>
         /// <param name="options">The options instance.</param>
         /// <returns>The <see cref="ValidateOptionsResult"/> result.</returns>
+        [RequiresUnreferencedCode("The type of options cannot be statically discovered.")]
```

Comment:
Can't we annotate `TOptions` as `[DAM(Properties)]` and then suppress these warnings?

I believe that DataAnnotations validation only validates the top-level properties of the class.

However, thinking more about this, if `TOptions` is a base type, and the `options` instance is a derived type, DataAnnotations validation will get the actual derived type's properties. So maybe doing that still wouldn't make this trim-safe.

Reply:
Yeah, I've added the DAM attribute to TOptions but I kept the RUC attribute in case base-sub class scenario.